### DISCUSSION TITLE
Automatically synthesizing help strings for enum valued flags

### DISCUSSION THREAD

**Discussion by jiawen** [abseil/abseil-cpp#1750] [Ideas]
The flags library has pretty good support for enum valued flags via the `AbslParseFlag` and `AbslUnparseFlag` extension points. In the course of maintaining a codebase that uses it, one thing that came up is divergence between the manually written help message and the set of enum values. I.e.,

```
ABSL_FLAG(my_ns::MyEnumType, my_flag,  my_ns::MyEnumType::kRed,
          "Permitted values: red, green, blue");
```

As you can imagine, if I add `kYellow` to `my_ns::MyEnumType` and corresponding support in `AbslParseFlag` and `AbslUnparseFlag`, the help message will not be updated. I have to grep the entire codebase for all uses of this enum as a flag and update the error messages.

I'd like to explore the design space for synthesizing the help message automatically. Ideally, it would take advantage of `AbslUnparseFlag` to iterate over all enum values (!) and return a string help message, potentially with a preamble and epilogue.

As I type this, I realize that the developer could of course manually do this as:

```
ABSL_FLAG(my_ns::MyEnumType, my_flag,  my_ns::MyEnumType::kRed,
          makeHelpMessage<my_ns::MyEnumType>("preamble", "epilogue"));  // With template magic to iterate over the enum values.
```

**Comment by derekmauro** [comment]
What you are asking for is enum reflection. Abseil won't be implementing that. https://github.com/Neargye/magic_enum is a library that I know of that does that.
